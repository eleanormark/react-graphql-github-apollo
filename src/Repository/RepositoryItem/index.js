import React from 'react';
import Link from '../../Link';

const RepositoryItem = ({
  name,
  url,
  descriptionHTML,
  primaryLanguage,
  owner,
  stargazers,

}) => (
  <div>
    <div className="RepositoryItem-title">

      <div className="RepositoryItem-title-url">
        <Link href={url}>{name}</Link>
      </div>

      <div className="RepositoryItem-title-action">
        {stargazers.totalCount} Stars
      </div>
    </div>

    <div className="RepositoryItem-description">
      <div
        className="RepositoryItem-description-info"
        dangerouslySetInnerHTML={{ __html: descriptionHTML }}
      />
      <div className="RepositoryItem-description-details">
        <div>
          {primaryLanguage && (
            <span>Language: {primaryLanguage.name}</span>
          )}
        </div>
        <div>
          {owner && (
            <span>
              Owner: <a href={owner.url}>{owner.login}</a>
            </span>
          )}
        </div>
      </div>
    </div>
  </div>
);

export default RepositoryItem;