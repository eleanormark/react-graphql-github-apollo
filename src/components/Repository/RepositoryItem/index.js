import React from "react";
import { Mutation } from "react-apollo";
import REPOSITORY_FRAGMENT from "../fragments";
import Link from "../../Link";
import Button from "../../Button";

import { STAR_REPOSITORY, UNSTAR_REPOSITORY } from "../mutations";

const updateAddStar = (
  client,
  {
    data: {
      addStar: {
        starrable: { id, viewerHasStarred }
      }
    }
  }
) =>
  client.writeFragment({
    id: `Repository:${id}`,
    fragment: REPOSITORY_FRAGMENT,
    data: getUpdatedStarData(client, id, viewerHasStarred)
  });

const updateRemoveStar = (
  client,
  {
    data: {
      removeStar: {
        starrable: { id, viewerHasStarred }
      }
    }
  }
) => {
  client.writeFragment({
    id: `Repository:${id}`,
    fragment: REPOSITORY_FRAGMENT,
    data: getUpdatedStarData(client, id, viewerHasStarred)
  });
};

const getUpdatedStarData = (client, id, viewerHasStarred) => {
  const repository = client.readFragment({
    id: `Repository:${id}`,
    fragment: REPOSITORY_FRAGMENT
  });

  let { totalCount } = repository.stargazers;
  totalCount = viewerHasStarred ? totalCount + 1 : totalCount - 1;

  return {
    ...repository,
    stargazers: {
      ...repository.stargazers,
      totalCount
    }
  };
};

const RepositoryItem = ({
  id,
  name,
  url,
  descriptionHTML,
  primaryLanguage,
  owner,
  stargazers,
  viewerHasStarred
}) => (
  <div>
    <div>
      <div>
        <Link href={url}>{name}</Link>
      </div>
    </div>
    <div>
      {!viewerHasStarred ? (
        <Mutation
          mutation={STAR_REPOSITORY}
          variables={{ id }}
          update={updateAddStar}
        >
          {(addStar, { data, loading, error }) => (
            <Button white onClick={addStar}>
              {stargazers.totalCount} Star
            </Button>
          )}
        </Mutation>
      ) : (
        <span>
          {
            <Mutation
              mutation={UNSTAR_REPOSITORY}
              variables={{ id }}
              update={updateRemoveStar}
            >
              {(removeStar, { data, loading, error }) => (
                <Button white onClick={removeStar}>
                  {stargazers.totalCount} Unstar
                </Button>
              )}
            </Mutation>
          }
        </span>
      )}
    </div>
    <div>
      <div dangerouslySetInnerHTML={{ __html: descriptionHTML }} />
      <div>
        <div>
          {primaryLanguage && <span>Language: {primaryLanguage.name}</span>}
        </div>
        <div>
          {owner && (
            <span>
              Owner: <a href={owner.url}>{owner.login}</a>
            </span>
          )}
        </div>
      </div>
    </div>
  </div>
);

export default RepositoryItem;
