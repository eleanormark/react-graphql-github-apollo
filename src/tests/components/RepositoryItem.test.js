import React from 'react';
import { shallow, mount} from 'enzyme';
import 'jest-styled-components';
import RepositoryItem from "../../components/Repository/RepositoryItem";
import repos from '../fixture/repos';

test("should render RepositoryItem correctly", () => {
    const wrapper = shallow(<RepositoryItem repos = {repos[0]}/>);
    // wrapper.find('Button').stimulate('click')
    // expect(wrapper.find('Button').text).toBe('1 Unstar');
    expect(wrapper).toMatchSnapshot();
});