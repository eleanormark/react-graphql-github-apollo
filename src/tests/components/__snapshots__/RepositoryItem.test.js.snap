// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RepositoryItem correctly 1`] = `
<div>
  <div>
    <div>
      <Link />
    </div>
  </div>
  <div>
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "input",
                        },
                        "value": Object {
                          "fields": Array [
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "starrableId",
                              },
                              "value": Object {
                                "kind": "Variable",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                              },
                            },
                          ],
                          "kind": "ObjectValue",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "addStar",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "starrable",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "selections": Array [
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "viewerHasStarred",
                                },
                                "selectionSet": undefined,
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "ID",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 139,
            "start": 0,
          },
        }
      }
      update={[Function]}
      variables={
        Object {
          "id": undefined,
        }
      }
    >
      <Component />
    </Mutation>
  </div>
  <div>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
    <div>
      <div />
      <div />
    </div>
  </div>
</div>
`;
